Fix169

janvanrijn opened the pull request.

janvanrijn commented on openml/runs/functions.py:
```diff
@@ -88,6 +94,24 @@ def run_task(task, model, avoid_duplicate_runs=True, flow_tags=None):
 
     return run
 
+def initialize_model_from_run(run_id):
```
Agreed, I added tests.

janvanrijn commented on openml/runs/functions.py:
```diff
@@ -111,6 +135,44 @@ def _run_exists(task_id, setup_id):
         return False
 
+def _get_seeded_model(model, seed=None):
+    '''Sets all the non-seeded components of a model with a seed.
```
I'm not sure. One could argue that that is a restriction of the run_tasks function, as that is the function the user interacts with. Furthermore, that function is responsible for the check.

This function only adds seeds to unseeded models

mfeurer replied:
But it can raise an exception:

```
import numpy as np
import sklearn.ensemble

import openml


rf = sklearn.ensemble.RandomForestClassifier(
    random_state=np.random.RandomState(1))
openml.runs.functions._get_seeded_model(rf, 5)
```
But you're right, it should be documented in the `run_tasks()` function.

janvanrijn replied:
Mea culpa, i will add it

janvanrijn commented on openml/runs/functions.py:
```diff
@@ -111,6 +135,44 @@ def _run_exists(task_id, setup_id):
+    for param_name in sorted(model_params):
+        if 'random_state' in param_name:
+            currentValue = model_params[param_name]
+            # important to draw the value at this point (and not in the if statement)
```
Added description. 

I'm not sure if we really need this, but seems nice property to respect.

mfeurer replied:
k

mfeurer commented on openml/setups/setup.py:
```diff
@@ -0,0 +1,35 @@
+
+       Parameters
+       ----------
+       flow_id : int
```
It looks like you copied the docstring from the object above and didn't adapt it.

janvanrijn replied:
I loove copy/pasting. fixed it.

mfeurer commented on openml/runs/run.py:
```diff
@@ -176,50 +176,38 @@ def _parse_parameters(model, flow):
+                    _current['oml:component'] = main_id
                 else:
-                    raise ValueError("parameter %s not in flow description of flow %s" %(param,flow.name))
+                    _current['oml:component'] = _param_dict[_flow.name]
```
Why is it once an ID, and once a name?

mfeurer commented on openml/runs/run.py:
```diff
@@ -165,7 +165,7 @@ def _create_description_xml(self):
 
     @staticmethod
-    def _parse_parameters(model, flow):
+    def _parse_parameters(model, server_flow):
```
Looking at this again I'm actually surprised that this is not called `run_task`, but only when publishing. But maybe this should be its own issue/PR.

janvanrijn replied:
Can you elaborate a bit? I don really understand which / why

mfeurer replied:
Sorry, I meant "that this is not called IN `run_task`"

mfeurer commented on tests/test_setups/test_setup_functions.py:
```diff
@@ -45,33 +46,47 @@ def test_nonexisting_setup_exists(self):
+            current = openml.setups.get_setup(setups[idx])
+            assert current.flow_id > 0
+            if num_params[idx] == 0:
+                assert current.parameters is None
```
Could you please use `self.assert` in the unit tests? It gives nicer outputs.

janvanrijn replied:
You mean self.asserts()?
Sure.

janvanrijn commented:
test_existing_setup_exists makes use of sentinels. That gives us a general problem when serializing a flow (or setup) and comparing it to one on the server, as the flow (setup) serialization is not aware of the (and which) sentinel string was used. For that reason, I slightly generalized one of the functions, such that it does not rely on name mappings for the main flow (boolean flag that indicates a function call on depth 1). This way, we do not have to remove the sentinels and we can in this context still test flows without subflows.

mfeurer commented on tests/test_setups/test_setup_functions.py:
```diff
@@ -87,6 +87,6 @@ def test_setup_get(self):
             assert current.flow_id > 0
             if num_params[idx] == 0:
-                self.asserts(current.parameters is None)
+                self.assertTrue(current.parameters is None)
```
Sorry, I should have been more specific. The correct one here would be `self.assertIsNone`.

mfeurer commented on tests/test_setups/test_setup_functions.py:
```diff
@@ -87,6 +87,6 @@ def test_setup_get(self):
+                self.assertTrue(current.parameters is None)
             else:
-                self.asserts(len(current.parameters) == num_params[idx])
+                self.assertTrue(len(current.parameters) == num_params[idx])
```
Sorry, I should have been more specific. The correct one here would be `self.assertEqual`.

mfeurer commented:
Thanks for the explanation of the sentinels. Maybe it would be good to add that to the actual unit test. What I'm wondering right now is if this is save with respect to running the unit tests in parallel (which happens on travis-ci). You assume:
```
# although the flow exists, we can be sure there are no
# setups (yet) as it hasn't been ran
```
it could happen that it was already run. Maybe adding a sentinel to a hyperparameter? Or is there some other way of making a setup unique?

Once we figured out this part, I think we can merge this PR before it becomes too big and create new PRs for the missing functionality.

janvanrijn commented:
I had already added this in the comment:
```
# because of the sentinel, we can not use flows that contain subflows
```

Also, the other comment was slightly confusing. I changed this to:
```
# although the flow exists (created as of previous statement),
# we can be sure there are no setups (yet) as it was just created
# and hasn't been ran
```

Should be fine now? Let's merge
